packages/lesswrong/server/tagging/tagCallbacks.js

```diff
@@ -0,0 +1,75 @@
+import { addCallback } from 'meteor/vulcan:core';
+import { Tags, tagDescriptionEditableOptions } from '../../lib/collections/tags/collection.js';
+import { TagRels } from '../../lib/collections/tagRels/collection.js';
+import { addEditableCallbacks } from '../editor/make_editable_callbacks.js'
+import Users from 'meteor/vulcan:users';
+import { performVoteServer } from '../voteServer.js';
+
+function isValidTagName(name) {
+  // Name must be nonempty and use only a restricted set of characters
+  if (!/^#?[a-zA-Z0-9-]+$/.test(name))
+    return false;
+  
+  return true;
+}
+
+function normalizeTagName(name) {
+  // If the name starts with a hash, strip it off
+  if (name.startsWith("#"))
+    return name.substr(1);
+  else
+    return name;
+}
+
+addCallback("tags.create.validate", ({ document: tag }) => {
+  if (!isValidTagName(tag.name))
+    throw new Error("Invalid tag name (use only letters, digits and dash)");
+  
+  // If the name starts with a hash, strip it off
+  const normalizedName = normalizeTagName(tag.name);
+  if (tag.name !== normalizedName) {
+    tag = {
+      ...tag,
+      name: normalizedName,
+    };
+  }
+  
+  // Name must be unique
+  const existing = Tags.find({name: normalizedName, deleted:false}).fetch();
```

Comment:
I'm somewhat surprised this infrastructure doesn't already exist in a more dry format. Not objecting to merging with it, just surprised.